File: src/kc_format.c
/* Dali's positive-integer interlaced Elias representation. */
#include <limits.h>
#include "kc_format.h"

static unsigned int kc_floor_log2(uint32_t value) {
#ifdef __GNUC__
   return 31U - (unsigned int)__builtin_clz(value);
#else
   unsigned int result = 0;
   while (value >>= 1)
      result++;
   return result;
#endif
}

unsigned int kc_dali_length_bit_cost(uint32_t value) {
   return value == 0 ? 0 : (kc_floor_log2(value) << 1) + 1;
}

typedef struct kc_length_writer {
   kc_put_bit_fn put_bit;
   void *context;
   int skip;
} kc_length_writer;

static int kc_emit_bit(kc_length_writer *writer, unsigned int bit) {
   if (writer->skip) {
      writer->skip = 0;
      return 1;
   }
   return writer->put_bit(writer->context, bit & 1U);
}

int kc_write_dali_length(uint32_t value, int skip_first_bit,
      kc_put_bit_fn put_bit, void *context) {
   kc_length_writer writer;
   unsigned int bits;
   uint32_t marker;

   if (value == 0 || put_bit == NULL)
      return 0;

   writer.put_bit = put_bit;
   writer.context = context;
   writer.skip = skip_first_bit != 0;
   bits = kc_floor_log2(value);
   marker = UINT32_C(1) << bits;
   if (bits >= 8) {
      value &= ~marker;
      value = (value >> 8) | ((value & UINT32_C(0xff)) << (bits - 8));
   }
   while ((marker >>= 1) != 0) {
      if (!kc_emit_bit(&writer, 0) ||
          !kc_emit_bit(&writer, (value & marker) != 0))
         return 0;
   }
   return kc_emit_bit(&writer, 1);
}

static int kc_read_one(kc_get_bit_fn get_bit, void *context, unsigned int *bit) {
   return get_bit != NULL && get_bit(context, bit);
}

int kc_read_dali_length(int first_zero_consumed, kc_get_bit_fn get_bit,
      void *context, uint32_t *result) {
   uint32_t value = 1;
   unsigned int control;
   unsigned int payload;
   unsigned int bits;

   if (get_bit == NULL || result == NULL)
      return 0;
   if (first_zero_consumed) {
      control = 0;
   }
   else if (!kc_read_one(get_bit, context, &control)) {
      return 0;
   }
   while (control == 0) {
      if (!kc_read_one(get_bit, context, &payload) || value > (UINT32_MAX >> 1))
         return 0;
      value = (value << 1) | payload;
      if (!kc_read_one(get_bit, context, &control))
         return 0;
   }
   bits = kc_floor_log2(value);
   if (bits >= 8) {
      uint32_t payload_mask = (UINT32_C(1) << bits) - 1;
      uint32_t encoded = value & payload_mask;
      uint32_t low_eight = encoded >> (bits - 8);
      uint32_t rest_mask = (UINT32_C(1) << (bits - 8)) - 1;
      uint32_t rest = encoded & rest_mask;
      value = (UINT32_C(1) << bits) | (rest << 8) | low_eight;
   }
   *result = value;
   return 1;
}

File: plugin/src/main/java/kabutocrunch/KcFormat.java
/*
 * Kabuto stream constants and Dali-derived Elias length helpers.
 *
 * Java port of src/kc_format.c. Behaviour is bit-exact with the C encoder;
 * see plugin/README.md and tools/test_plugin.py.
 */
package kabutocrunch;

final class KcFormat {

    public static final int MIN_OFFSET = 1;
    public static final int MAX_OFFSET = 0x7f80;
    public static final int MAX_VARLEN = 0xffff;
    public static final int BLOCK_SIZE = 0x10000;
    public static final int MIN_MATCH_SIZE = 1;

    private KcFormat() {
    }

    /** Interface mirroring kc_put_bit_fn: returns 0 to signal failure. */
    public interface PutBit {
        int putBit(int bit);
    }

    /** Interface mirroring kc_get_bit_fn: returns 0 to signal failure. */
    public interface GetBit {
        int getBit(int[] bit);
    }

    /** floor(log2(value)); caller must not pass 0. */
    public static int floorLog2(int value) {
        return 31 - Integer.numberOfLeadingZeros(value);
    }

    public static int daliLengthBitCost(int value) {
        if (value == 0) {
            return 0;
        }
        return (floorLog2(value) << 1) + 1;
    }

    private static final class LengthWriter {
        final PutBit putBit;
        int skip;

        LengthWriter(PutBit putBit) {
            this.putBit = putBit;
        }
    }

    private static int emitBit(LengthWriter writer, int bit) {
        if (writer.skip != 0) {
            writer.skip = 0;
            return 1;
        }
        return writer.putBit.putBit(bit & 1);
    }

    public static int writeDaliLength(int value, int skipFirstBit, PutBit putBit) {
        if (value == 0 || putBit == null) {
            return 0;
        }

        LengthWriter writer = new LengthWriter(putBit);
        writer.skip = skipFirstBit != 0 ? 1 : 0;

        int bits = floorLog2(value);
        int marker = 1 << bits;
        if (bits >= 8) {
            value &= ~marker;
            value = (value >>> 8) | ((value & 0xff) << (bits - 8));
        }
        while ((marker >>>= 1) != 0) {
            if (emitBit(writer, 0) == 0 || emitBit(writer, (value & marker) != 0 ? 1 : 0) == 0) {
                return 0;
            }
        }
        return emitBit(writer, 1);
    }

    private static boolean readOne(GetBit getBit, int[] bit) {
        return getBit != null && getBit.getBit(bit) != 0;
    }

    public static boolean readDaliLength(int firstZeroConsumed, GetBit getBit, int[] result) {
        if (getBit == null || result == null) {
            return false;
        }
        int value = 1;
        int control;
        int payload;
        int bits;

        int[] box = new int[1];
        if (firstZeroConsumed != 0) {
            control = 0;
        } else {
            if (!readOne(getBit, box)) {
                return false;
            }
            control = box[0];
        }
        while (control == 0) {
            if (!readOne(getBit, box)) {
                return false;
            }
            payload = box[0];
            if (value > (Integer.MAX_VALUE >>> 1)) {
                return false;
            }
            value = (value << 1) | payload;
            if (!readOne(getBit, box)) {
                return false;
            }
            control = box[0];
        }
        bits = floorLog2(value);
        if (bits >= 8) {
            int payloadMask = (1 << bits) - 1;
            int encoded = value & payloadMask;
            int lowEight = encoded >>> (bits - 8);
            int restMask = (1 << (bits - 8)) - 1;
            int rest = encoded & restMask;
            value = (1 << bits) | (rest << 8) | lowEight;
        }
        result[0] = value;
        return true;
    }
}
